File: Data Structures & Algorithms/daily-temperatures/submission-2.java
class Solution {
    public int[] dailyTemperatures(int[] temperatures) {
        int[] res = new int[temperatures.length];
        Stack<int[]> s = new Stack<>();

        for (int i = 0; i < temperatures.length; i++) {
            int temp = temperatures[i];
            while (!s.isEmpty() && s.peek()[0] < temp) {
                int[] popped = s.pop();
                res[popped[1]] = i - popped[1];
            }
            s.add(new int[]{temp, i});
        }

        return res;
    }
}

File: Data Structures & Algorithms/largest-rectangle-in-histogram/submission-2.py
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        res = 0
        stack = []

        for i in range(len(heights)):
            start = i
            while stack and stack[-1][0] > heights[i]:
                height, index = stack.pop()
                res = max(res, height * (i - index))
                start = index
            stack.append((heights[i], start))

        for h, i in stack:
            res = max(res, h * (len(heights) - i))

        return res